File: ctci/java/ch1-arrays-and-strings/q1_is_unique/QuestionA.java
package q1_is_unique;

import java.util.HashSet;

/**
 * Created by jjlee on 7/2/16.
 */
public class QuestionA {
    /* assume ASCII */
    public static boolean isUnique(String s) {
        if (s.length() > 128)
            return false;

        boolean[] exists = new boolean[128];
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (exists[c])
                return false;
            exists[c] = true;
        }
        return true;
    }

    /* no assumption */
    public static boolean isUniqueNoAssumption(String s) {
        HashSet<Character> exists = new HashSet<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (exists.contains(c))
                return false;
            exists.add(c);
        }
        return true;
    }
}

File: ctci/python/ch1-arrays-and-strings/q1-is-unique/QuestionA.py
# assume ASCII
def is_unique(s):
    if len(s) > 128:
        return False

    exists = [False for _ in xrange(128)]
    for c in s:
        if exists[ord(c)]:
            return False
        exists[ord(c)] = True
    return True


# no assumption
def is_unique_no_assumption(s):
    exists = set()
    for c in s:
        if c in exists:
            return False
        exists.add(c)
    return True
